homeassistant/components/ipp_printer/sensor.py

```diff
@@ -0,0 +1,190 @@
+                                        .index(marker)))
+
+    add_entities(dev, True)
+
+
+class MarkerSensor(Entity):
+    """Implementation of the MarkerSensor.
+
+    This sensor represents the percentage of ink or toner.
+    """
+
+    def __init__(self, data, name, index):
+        """Initialize the sensor."""
+        self.data = data
+        self._name = name
+        self._index = index
+        self._attributes = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        return ICON_MARKER
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self._attributes is not None:
+            return self._attributes['marker-levels'][self._index]
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return "%"
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        if self._attributes is not None:
+            high_level = self._attributes['marker-high-levels']
+            if isinstance(high_level, list):
+                high_level = high_level[self._index]
+
+            low_level = self._attributes['marker-low-levels']
+            if isinstance(low_level, list):
+                low_level = low_level[self._index]
+
+            marker_types = self._attributes['marker-types']
+            if isinstance(marker_types, list):
+                marker_types = marker_types[self._index]
+
+            return {
+                ATTR_MARKER_HIGH_LEVEL: high_level,
+                ATTR_MARKER_LOW_LEVEL: low_level,
+                ATTR_MARKER_TYPE: marker_types,
+                ATTR_PRINTER_NAME: self._attributes['printer-make-and-model']
+            }
+
+    def update(self):
+        """Update the state of the sensor."""
+        # Data fetching is done by PrinterSensor
+        self._attributes = self.data.attributes
+
+
+class PrinterSensor(Entity):
+    """Implementation of the PrinterSensor.
+
+    This sensor represents the status of the printer.
+    """
+
+    def __init__(self, data, name):
+        """Initialize the sensor."""
+        self.data = data
+        self._name = name
+        self._status = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        return ICON_PRINTER
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self._status is not None:
+            return self._status
+
+    def update(self):
+        """Fetch new state data for the sensor."""
+        try:
+            self.data.update()
+            status = self.data.attributes["printer-state"]
```

Comment:
Do a dict lookup here, [similar to this](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/cups/sensor.py#L86).